packages/stan/src/derived.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+
+const resolveQueue = createQueue();
+
+/**
+ * Creates a derived atom.
+ *
+ * @template T
+ * @param {import('./types').WPDerivedAtomResolver<T>} resolver Atom Resolver.
+ * @param {import('./types').WPDerivedAtomUpdater}     updater  Atom updater.
+ * @param {boolean=}                                   isAsync  Atom resolution strategy.
+ * @param {string=}                                    id       Atom id.
+ * @return {import("./types").WPAtom<T>} Createtd atom.
+ */
+
+export const createDerivedAtom = (
+	resolver,
+	updater = noop,
+	isAsync = false,
+	id
+) => ( registry ) => {
+	/**
+	 * @type {any}
+	 */
+	let value = null;
+
+	/**
+	 * @type {(() => void)[]}
+	 */
+	let listeners = [];
+
+	/**
+	 * @type {(import("./types").WPAtomState<any>)[]}
+	 */
+	let dependencies = [];
+	let isListening = false;
+	let isResolved = false;
+	const context = {};
+
+	const dependenciesUnsubscribeMap = new WeakMap();
+
+	const notifyListeners = () => {
+		listeners.forEach( ( l ) => l() );
+	};
+
+	const refresh = () => {
+		if ( listeners.length ) {
+			if ( isAsync ) {
+				resolveQueue.add( context, resolve );
+			} else {
+				resolve();
+			}
+		} else {
+			isListening = false;
+		}
+	};
+
+	/**
+	 * @param {import('./types').WPAtomState<any>} atomState
+	 */
+	const addDependency = ( atomState ) => {
+		if ( ! dependenciesUnsubscribeMap.has( atomState ) ) {
+			dependenciesUnsubscribeMap.set(
+				atomState,
+				atomState.subscribe( refresh )
```

Comment:
One optimization idea, not sure how impactful: currently, all derived atoms maintain subscriptions to their dependencies, even if their own consumers unsubscribed and are not longer interested in the value.

The derived atom could unsubscribe from dependencies once its own subscriber count drops to zero. Two scenarios to watch for:
- when calling `.get()`, ensure we're not returning a stale value
- there is a "dependency changed while we were resolving so we need to rerun the resolver" case that might be affected

Reply:
I talked to @grzim today and he noted how utilizing Rx.js could drastically reduce the amount of code here - I know he'll want to follow-up with more details but I wanted to add this idea to the discussion early.

Reply:
My first thought here is that stan is approximatively 1.5kb now and I don't want to add Rx.js just for that. Thought it's been a long time I didn't look at that, maybe we can reuse a very small part of it.

Reply:
Besides RxJS, there's also a [proposal to add `Observable` directly to the language](https://github.com/tc39/proposal-observable). There's one implementation that's directly in `core-js` and multiple other ones with various levels of minimalism (or maximalism). Even RxJS should be compatible with it.

Reply:
I'm playing with it right now just to assess the impact on bundle size, I'll update this thread soon.

Reply:
With a few Rx.js operators imported AND webpack tree shaking enabled, the size of stan goes up to 20-25kb, or 6.4kb gzipped (vs 4.9K / 1.7K gzipped before). It's not terrible, but I stumbled upon a potential roadblock. `stan` is really simple and most of the code would be necessary even if we relied on Observables, e.g. I don't see how tracking dependencies in `derived.js` could be simplified. Perhaps this library is small enough that using functional programming while exposing an imperative API is counter-productive? I imagine it would make more sense if the public API exposed observables, but this increases the barrier of entry.

Reply:
By returning observables from derived atoms, I mean something like:

```js
const count1 = createAtom( 1 );
const count2 = createAtom( 1 );
const count3 = createAtom( 1 );
const sum = createDerivedAtom( ( { get } ) =>
	combineLatest( get( count1 ), get( count2 ), get( count3 ) )
		.pipe( map( ( [ a, b, c ] ) => a + b + c ) )
);
```

In which case, `createDerivedAtom` could be as simple as:

```js
export const createDerivedAtom = ( resolver, updater = noop, id ) => (
	registry
) => {
	let lastValue = null;
	const get = ( atom ) => registry.__unstableGetAtomState( atom ).subject;
	const observable = resolver( { get } ).pipe(
		tap( ( value ) => ( lastValue = value ) )
	);
	return {
		id,
		type: 'derived',
		get: () => lastValue,
		subscribe: ( listener ) => observable.subscribe( listener ),
		async set( action ) {
			await updater(
				{
					get: ( atom ) => registry.get( atom ),
					set: ( atom, arg ) => registry.set( atom, arg ),
				},
				action
			);
		},
	}
```